### ISSUE TITLE
System::arrayCopy attached to the wrong caller

### ISSUE THREAD

**Issue by franz1981** [jvm-profiling-tools/async-profiler#567]
Code reproducer:
```java
@State(Scope.Benchmark)
@BenchmarkMode(Mode.Throughput)
@OutputTimeUnit(TimeUnit.SECONDS)
@Warmup(iterations = 10, time = 1)
@Measurement(iterations = 20, time = 1)
@Fork(1)
public class ProfilerFallacy {

    @Param("1000")
    private int work;
    private byte[] bytes;
    private byte v;

    @Setup
    public final void setup() {
        bytes = new byte[work];
    }

    private void blameMePlease() {
        notInlineableCall();
    }

    @CompilerControl(CompilerControl.Mode.DONT_INLINE)
    private void notInlineableCall() {

    }

    @Benchmark
    @CompilerControl(CompilerControl.Mode.DONT_INLINE)
    public void doIntrinsicAndBlameThePollReturnOwner() {
        System.arraycopy(bytes, 0, bytes, 0, bytes.length);
        blameMePlease();
    }

}
```
on async profiler (adding debug non safepoints):
![async-profiler](https://user-images.githubusercontent.com/13125299/159523401-62409027-d918-4f4c-996f-4e2a5b9ebac1.png)

on perf (adding preserve frame point and debug non safepoints):
![perf](https://user-images.githubusercontent.com/13125299/159523451-431cd1c2-8fa2-4f89-a61b-66bd2eee9845.png)

It happens both on JDK 11 (ie JDK 11.0.14.1, OpenJDK 64-Bit Server VM, 11.0.14.1+1) and JDK 14/17
async-profiler's stack trace report `blameMePlease` as caller of System::arrayCopy

`-prof perfasm` shows:
```
....[Hottest Region 2]..............................................................................
c2, level 4, red.hat.puzzles.profilers.ProfilerFallacy::doIntrinsicAndBlameThePollReturnOwner, version 494 (94 bytes) 

            0x00007fbb44339844: shl    $0x3,%r10
            0x00007fbb44339848: movabs $0x800000000,%r12
            0x00007fbb44339852: add    %r12,%r10
            0x00007fbb44339855: xor    %r12,%r12
            0x00007fbb44339858: cmp    %r10,%rax
            0x00007fbb4433985b: jne    0x00007fbb3c88b080  ;   {runtime_call ic_miss_stub}
            0x00007fbb44339861: data16 xchg %ax,%ax
            0x00007fbb44339864: nopl   0x0(%rax,%rax,1)
            0x00007fbb4433986c: data16 data16 xchg %ax,%ax
          [Verified Entry Point]
  0.15%     0x00007fbb44339870: mov    %eax,-0x14000(%rsp)
  1.29%     0x00007fbb44339877: push   %rbp
  0.20%     0x00007fbb44339878: mov    %rsp,%rbp
            0x00007fbb4433987b: sub    $0x10,%rsp         ;*synchronization entry
                                                          ; - red.hat.puzzles.profilers.ProfilerFallacy::doIntrinsicAndBlameThePollReturnOwner@-1 (line 57)
  1.34%     0x00007fbb4433987f: mov    %rsi,%rbx
  0.26%     0x00007fbb44339882: mov    0x14(%rsi),%r11d   ;*getfield bytes {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - red.hat.puzzles.profilers.ProfilerFallacy::doIntrinsicAndBlameThePollReturnOwner@1 (line 57)
            0x00007fbb44339886: mov    0xc(%r12,%r11,8),%r10d  ;*arraylength {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - red.hat.puzzles.profilers.ProfilerFallacy::doIntrinsicAndBlameThePollReturnOwner@14 (line 57)
                                                          ; implicit exception: dispatches to 0x00007fbb443398df
  0.01%     0x00007fbb4433988b: test   %r10d,%r10d
         ╭  0x00007fbb4433988e: jbe    0x00007fbb443398af
  1.22%  │  0x00007fbb44339890: lea    (%r12,%r11,8),%r8  ;*getfield bytes {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - red.hat.puzzles.profilers.ProfilerFallacy::doIntrinsicAndBlameThePollReturnOwner@1 (line 57)
  0.17%  │  0x00007fbb44339894: lea    0x10(%r12,%r11,8),%rdi
         │  0x00007fbb44339899: movslq %r10d,%rdx
         │  0x00007fbb4433989c: mov    %rdi,%rsi
  1.46%  │  0x00007fbb4433989f: vzeroupper 
  1.59%  │  0x00007fbb443398a2: movabs $0x7fbb3c88c0e0,%r10
  0.19%  │  0x00007fbb443398ac: callq  *%r10              ;*invokestatic arraycopy {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - red.hat.puzzles.profilers.ProfilerFallacy::doIntrinsicAndBlameThePollReturnOwner@15 (line 57)
  0.01%  ↘  0x00007fbb443398af: mov    %rbx,%rsi
            0x00007fbb443398b2: xchg   %ax,%ax
            0x00007fbb443398b4: vzeroupper 
  1.63%     0x00007fbb443398b7: callq  0x00007fbb3ce8d400  ; ImmutableOopMap{}
                                                          ;*invokespecial notInlineableCall {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - red.hat.puzzles.profilers.ProfilerFallacy::blameMePlease@1 (line 28)
                                                          ; - red.hat.puzzles.profilers.ProfilerFallacy::doIntrinsicAndBlameThePollReturnOwner@19 (line 59)
                                                          ;   {optimized virtual_call}
            0x00007fbb443398bc: vzeroupper 
  1.58%     0x00007fbb443398bf: add    $0x10,%rsp
  1.45%     0x00007fbb443398c3: pop    %rbp
            0x00007fbb443398c4: mov    0x108(%r15),%r10
            0x00007fbb443398cb: test   %eax,(%r10)        ;   {poll_return}
  1.32%     0x00007fbb443398ce: retq                      ;*invokespecial notInlineableCall {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - red.hat.puzzles.profilers.ProfilerFallacy::blameMePlease@1 (line 28)
                                                          ; - red.hat.puzzles.profilers.ProfilerFallacy::doIntrinsicAndBlameThePollReturnOwner@19 (line 59)
            0x00007fbb443398cf: mov    %rax,%rsi
            0x00007fbb443398d2: vzeroupper 
            0x00007fbb443398d5: add    $0x10,%rsp
            0x00007fbb443398d9: pop    %rbp
            0x00007fbb443398da: jmpq   0x00007fbb3c93d400  ;   {runtime_call _rethrow_Java}
            0x00007fbb443398df: mov    $0xfffffff6,%esi
            0x00007fbb443398e4: vzeroupper 
            0x00007fbb443398e7: callq  0x00007fbb3c889e00  ; ImmutableOopMap{}
....................................................................................................
 13.88%  <total for region 2>
```
Meaning that 
```
  1.59%  │  0x00007fbb443398a2: movabs $0x7fbb3c88c0e0,%r10
  0.19%  │  0x00007fbb443398ac: callq  *%r10              ;*invokestatic arraycopy {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - red.hat.puzzles.profilers.ProfilerFallacy::doIntrinsicAndBlameThePollReturnOwner@15 (line 57)
```
calling to `StubRoutines::jbyte_disjoint_arraycopy` is correctly assigned to the right BCI.

NOTE: line numbers won't match because the original source code was bigger, but the point is: perfasm doesn't show BCI=-1

**Comment by franz1981** [comment]
Worth mentioning https://github.com/jvm-profiling-tools/async-profiler/issues/127 although it doesn't seem the same issue

**Comment by franz1981** [comment]
In addition, I've verified that BCI is correct according to javap: 

![image](https://user-images.githubusercontent.com/13125299/159576223-e12cc62d-5750-4680-a550-0b22a5a1eef4.png)

**Comment by apangin** [comment]
Thank you for the report and the test case.
How does the graph change if you add `safemode=4` profiler option?

**Comment by franz1981** [comment]
@apangin 

![image](https://user-images.githubusercontent.com/13125299/159593307-49bcb250-d781-4eea-b8be-3dde3ba7cda2.png)

And it's working as expected, great; hence it was due to some correction during stack-walking?

**Comment by apangin** [comment]
When I unwind the top frame manually, I set PC to the return address. However, the return address is the *next* instruction after the call, which may be attributed to a different bytecode.

Will check how to fix it the right way.

**Comment by aalexand** [comment]
Subtract 1 from the address maybe?

**Comment by apangin** [comment]
@aalexand Yes, more precisely, `sizeof(instruction_t)`. This woks for normal `call` instructions, but I believe I've seen cases where the return address was pushed onto the stack manually.

**Comment by apangin** [comment]
Fixed in release 2.8
